### PR TITLE
change resetKinesisVideoConnection to terminate all active upload han…

### PR DESCRIPTION
**Pull request by chehefen**
…dles. Add unit test simulating curl timeout. When an upload get an error, only indicate rollback if there is no other active upload handle.

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: kinesis-video-pic/src/client/src/StreamEvent.c [hunk 1]
```diff
@@ -664,18 +662,60 @@ STATUS streamTerminatedEvent(PKinesisVideoStream pKinesisVideoStream, UPLOAD_HAN
...
+                    CHK_STATUS(stackQueueGetCount(pKinesisVideoStream->pUploadInfoQueue, &sessionCount));
+
+                    for (i = 0; i < sessionCount && !otherActiveSessionPresent; i++) {
+                        CHK_STATUS(stackQueueGetAt(pKinesisVideoStream->pUploadInfoQueue, i, &item));
+                        pUploadHandleInfoItem = (PUploadHandleInfo) item;
+                        CHK(pUploadHandleInfoItem != NULL, STATUS_INTERNAL_ERROR);
+
+                        if (pUploadHandleInfoItem != pUploadHandleInfo &&
+                            (pUploadHandleInfoItem->state & UPLOAD_HANDLE_STATE_ACTIVE) != UPLOAD_HANDLE_STATE_NONE) {
+                            otherActiveSessionPresent = TRUE;
```
Comment (MushMal) [inline kinesis-video-pic/src/client/src/StreamEvent.c]: Perhaps describe here that we have the case of an old handle terminating but a new handle that's already streaming and in that case we leave it alone.

File: kinesis-video-pic/src/client/src/StreamEvent.c [hunk 2]
```diff
@@ -684,20 +724,22 @@ STATUS streamTerminatedEvent(PKinesisVideoStream pKinesisVideoStream, UPLOAD_HAN
...
 
-    // Stop the stream
-    pKinesisVideoStream->streamState = STREAM_STATE_STOPPED;
+    if (spawnNewUploadSession) {
+        // Stop the stream
+        pKinesisVideoStream->streamState = STREAM_STATE_STOPPED;
 
-    // Get the accepted state
-    CHK_STATUS(getStateMachineState(pKinesisVideoStream->base.pStateMachine, STREAM_STATE_STOPPED, &pState));
+        // Get the accepted state
```
Comment (MushMal) [inline kinesis-video-pic/src/client/src/StreamEvent.c]: Seems that we have \t in the lines here and below

File: kinesis-video-pic/src/client/src/StreamEvent.c (hunk 1 above)
Comment (MushMal) [reply kinesis-video-pic/src/client/src/StreamEvent.c]: One more thing. We should explicitly do something when we get into the state of a last ACK awaiting upload handle being dropped without durable persistence. Perhaps it's not at this layer but in C SDK layer. As we discussed, calling the error callback might be one way for now until we add a filter or have separate callback for fatal/unrecoverable errors. At least we should log.

File: kinesis-video-pic/src/client/src/StreamEvent.c [hunk 3]
```diff
@@ -686,9 +686,23 @@ STATUS streamTerminatedEvent(PKinesisVideoStream pKinesisVideoStream, UPLOAD_HAN
                             (pUploadHandleInfoItem->state & UPLOAD_HANDLE_STATE_ACTIVE) != UPLOAD_HANDLE_STATE_NONE) {
                             otherActiveSessionPresent = TRUE;
                             pKinesisVideoStream->connectionState = UPLOAD_CONNECTION_STATE_NOT_IN_USE;
+                            // dont spawn new session since there is already an active one.
                             spawnNewUploadSession = FALSE;
-                        }
 
+                            DLOGW("Upload handle %" PRIu64 " terminated with result %u while other upload handles are active."
+                                "Data sent through upload handle %" PRIu64 " may not be fully persisted.",
```
Comment (unicornss) [inline kinesis-video-pic/src/client/src/StreamEvent.c]: Are we referring "data sent through" as one fragment size bytes or can it be multiple fragments?

File: kinesis-video-pic/src/client/include/com/amazonaws/kinesis/video/client/Include.h [hunk 4]
```diff
@@ -185,6 +185,7 @@ extern "C" {
 #define STATUS_MAX_FRAME_TIMESTAMP_DELTA_BETWEEN_TRACKS_EXCEEDED                    STATUS_CLIENT_BASE + 0x00000085
 #define STATUS_STREAM_SHUTTING_DOWN                                                 STATUS_CLIENT_BASE + 0x00000086
 #define STATUS_CLIENT_SHUTTING_DOWN                                                 STATUS_CLIENT_BASE + 0x00000087
+#define STATUS_PUTMEDIA_LAST_PERSIST_ACK_NOT_RECEIVED								STATUS_CLIENT_BASE + 0x00000088
```
Comment (unicornss) [inline kinesis-video-pic/src/client/include/com/amazonaws/kinesis/video/client/Include.h]: wondering how this  status could help the user in terms of :
a) there is merely a logging status stating that we could not get the ACK within the ACK wait interval time but connection has been reset ? or
b) can the user retry the last fragment by cancelling the reset?

Comment (chehefen) [reply kinesis-video-pic/src/client/include/com/amazonaws/kinesis/video/client/Include.h]: Right now this is serving as a logging status. There is not much user can do to this status because the buffer may already be gone.

File: kinesis-video-pic/src/client/src/Stream.c [hunk 5]
```diff
@@ -1499,14 +1499,6 @@ STATUS getStreamData(PKinesisVideoStream pKinesisVideoStream, UPLOAD_HANDLE uplo
         deleteStreamUploadInfo(pKinesisVideoStream, pUploadHandleInfo);
         pUploadHandleInfo = NULL;
         pUploadHandleInfo = getStreamUploadInfoWithState(pKinesisVideoStream, UPLOAD_HANDLE_STATE_ACTIVE);
-
```
[the remark this answers is not in the document]
Comment (chehefen) [reply kinesis-video-pic/src/client/src/Stream.c]: Just realized that I shouldn't remove this clause. Putting it back and add comment. It's self-priming in case a upload handle is terminated and there is no other active upload handle to take over.

File: kinesis-video-pic/src/client/src/StreamEvent.c (hunk 2 above)
Comment (chehefen) [reply kinesis-video-pic/src/client/src/StreamEvent.c]: this block is inside a if statement

File: kinesis-video-pic/src/client/src/StreamEvent.c (hunk 3 above)
Comment (chehefen) [reply kinesis-video-pic/src/client/src/StreamEvent.c]: in most of the cases, it should be one fragment.

File: kinesis-video-pic/src/client/src/Stream.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (MushMal) [reply kinesis-video-pic/src/client/src/Stream.c]: Yeah, that's what I deduced. Could you remove LN 1503 then?

File: kinesis-video-pic/src/client/src/StreamEvent.c (hunk 3 above)
Comment (MushMal) [reply kinesis-video-pic/src/client/src/StreamEvent.c]: Thanks for adding this. Suggest rewording to stating that the "last fragment for upload handle blah might not be fully persisted". Also, output the fragment timestamp like when calling the error callback

File: kinesis-video-pic/src/client/src/Stream.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (chehefen) [reply kinesis-video-pic/src/client/src/Stream.c]: done

[the remark this answers is not in the document]
Comment (chehefen) [reply kinesis-video-pic/src/client/src/Stream.c]: done
